internal/mempool/mempool.go

```diff
@@ -654,40 +654,38 @@ func (mp *TxPool) hasMempoolInput(tx *dcrutil.Tx) bool {
 	return false
 }
 
-// fetchRedeemers returns all transactions that reference an outpoint for
-// the provided regular transaction `tx`. Returns nil if a non-regular
-// transaction is provided.
+// fetchRedeemers returns a set of transactions of type txType from the provided
+// pool that have an input referencing the provided regular transaction.
 //
 // This function MUST be called with the mempool lock held (for reads).
-func (mp *TxPool) fetchRedeemers(outpoints map[wire.OutPoint]*dcrutil.Tx, tx *dcrutil.Tx) []*dcrutil.Tx {
-	txType := stake.DetermineTxType(tx.MsgTx())
-	if txType != stake.TxTypeRegular {
-		return nil
-	}
-
-	tree := wire.TxTreeRegular
+func fetchRedeemers(tx *dcrutil.Tx, outpoints map[wire.OutPoint]*dcrutil.Tx,
```

Comment:
I'm really not a huge fan of this.  I know it was already there, but it's super wasteful to create allocations just do what is effectively filtering.

It would be cheaper to pass in a filter func instead of allocating and returning a slice.  Then there would only be a single small alloc for the closure, versus potentially multiple to build the slice.

Something like:

```Go
// Comments...
func forEachRedeemer(tx *dcrutil.Tx, outpoints map[wire.OutPoint]*dcrutil.Tx,
	pool map[chainhash.Hash]*TxDesc, txType stake.TxType, f func(redeemer *TxDesc) {

	...
		seen[*txRedeemer.Hash()] = struct{}{}
		f(redeemerTxDesc)
	}
	...
}
```

Also, because the argument list is quite long, I would suggest creating two separate funcs for the two cases:

```Go
// Comments...
func (mp *TxPool) forEachRedeemer(tx *dcrutil.Tx, f func(redeemer *TxDesc)) {
	forEachRedeemer(tx, mp.outpoints, mp.pool, f)
}
// Comments...
func (mp *TxPool) forEachStagedRedeemer(tx *dcrutil.Tx, f func(redeemer *TxDesc)) {
	forEachRedeemer(tx, mp.stagedOutpoints, mp.staged, f)
}
```

Then the callsites look something like the following, which I find more readable than the larger argument lists:

```Go
mp.forEachStagedRedeemer(tx, func(redeemer *TxDesc) {
	if redeemer.TxType != stake.TxTypeSStx {
		return
	}
	if mp.hasMempoolInput(redeemer.Tx) {
		return
	}
	...
})

...

mp.forEachRedeemer(tx, func(redeemer *TxDesc) {
	if redeemer.TxType != stake.TxTypeSStx {
		return
	}

	mp.removeTransaction(redeemer.Tx, true)
	...
})

Reply:
Makes sense, thank you. PR updated.